dev/mcal/src/Adc.cpp

```diff
@@ -0,0 +1,157 @@
+/**
+ * @file Adc.cpp
+ * @brief 
+ * @version 0.1
+ * @date 2024-07-20
+ * 
+ * @copyright Copyright (c) 2024
+ * 
+ */
+#include "mcal/inc/stm32f103xx.h"
+#include "Types.h"
+#include "Assert.h"
+#include "BitManipulation.h"
+#include "Rcc.h"
+#include "Adc.h"
+
+using namespace stm32::utils::bit_manipulation;
+using namespace stm32::registers::rcc; 
+using namespace stm32::dev::mcal::adc;
+using namespace stm32::registers::adc;
+
+
+ADC::ADC(const ADCConfig& config) : config_(config) {
+    switch (config_.number) {
+        case kADC1 : ADC_reg = (reinterpret_cast<volatile ADCRegDef*>(ADC1));
+                     break;
+        case kADC2 : ADC_reg = (reinterpret_cast<volatile ADCRegDef*>(ADC2));
+                     break;
+        default    : break;
+    }
+}
+
+void ADC::Init() {
+    ADC_reg->CR2.registerVal = 0;  // Reset ADC
+    ADC_reg->CR2.ALIGN = static_cast<RegWidth_t>(config_.alignment);
+    if (config_.trigSource != kSOFTWARE) {
+        if (config_.mode == kINJECTED) {
+            ADC_reg->CR2.JEXTTRIG = 1;
+            ADC_reg->CR2.JEXTSEL = static_cast<RegWidth_t>(config_.trigSource);
+        } else {
+            ADC_reg->CR2.EXTTRIG = 1;
+            ADC_reg->CR2.EXTSEL = static_cast<RegWidth_t>(config_.trigSource);
+        }
+    }
+    ADC_reg->CR2.ADON = 1;
+}
```

Comment:
To avoid multiple nested if we can rewrite the code:
```
   if (config_.trigSource == kSOFTWARE) {
       return;
    }
        if (config_.mode == kINJECTED) {
            ADC_reg->CR2.JEXTTRIG = 1;
            ADC_reg->CR2.JEXTSEL = static_cast<RegWidth_t>(config_.trigSource);
        } else {
            ADC_reg->CR2.EXTTRIG = 1;
            ADC_reg->CR2.EXTSEL = static_cast<RegWidth_t>(config_.trigSource);
        }
    ADC_reg->CR2.ADON = 1;
}
```